# <span style="color:turquoise">D208 Performance Assessment NBM2 Task 2</span>
## <span style="color:turquoise">Logistic Regression for Predictive Modeling</span>
&emsp;Ryan L. Buchanan
<br>&emsp;Student ID:  001826691
<br>&emsp;Masters Data Analytics (12/01/2020)
<br>&emsp;Program Mentor:  Dan Estes
<br>&emsp;(385) 432-9281 (MST)
<br>&emsp;rbuch49@wgu.edu
</span>

### <span style="color:green"><b>A1. Research Question</b>:</span>
Can we determine which individual customers are at high risk of churn? &nbsp; And, can we determine which features are most significant to churn?

### <span style="color:green"><b>A2. Objectives & Goals</b>:</span>
&emsp; Stakeholders in the company will benefit by knowing, with some measure of confidence, which customers are likely to churn soon. &nbsp; This knowledge will provide weight for decisions in marketing improved services to customers with these characteristics & past user experiences.

### <span style="color:green"><b>B1. Summary of Assumptions</b>:</span>
Assumptions of a logistic regression model include:
* It is based on Bernoulli (also, Binomial or Boolean) Distribution rather than Gaussian because the dependent variable is binary (in our dataset, to churn or not to churn).
* The predicted values are restricted to a range of nomial values: "Yes" or "No."
* It predicts the probability of a particular outcome rather than the outcome itself.
* There are no high correlations (multicollinearity) among predictors.
* It is the logarithm of the odds of achieving 1. &nbsp; In other words, a regression model, where the output is natural logarithm of the odds, also known as 'logit'.

### <span style="color:green"><b>B2. Tool Benefits</b>:</span>
&emsp; Python & IPython Jupyter notebooks will be used to support this analysis. &nbsp; Python offers an intuitive, simple & versatile programming style & syntax, as well as a large system of mature packages for data science & machine learning. &nbsp; Since, Python is cross-platform, it will work well whether consumers of the analysis are using Windows PCs or a MacBook laptop. &nbsp; It is fast when compared with other possible programming languages like R or MATLAB <span style="color:red">(Massaron, p. 8)</span>.
<br> &emsp; Also, there is strong support for Python as the most popular data science programming language in popular literature & media <span style="color:red">(CBTNuggets, p. 1)</span>.

### <span style="color:green"><b>B3. Appropriate Technique</b>:</span>
&emsp; Logistic regression is an appropriate technique to analyze the research question because or dependent variable is binomial, Yes or No. &nbsp; We want to find out what the likelihood of customer churn is for individual customers, based on a list of independent variables (area type, job, children, age, income, etc.). &nbsp; It will improve our understanding of increased probability of churn as we include or remove different independent variables & find out whether or not they have a positive or negative relationship to our target variable.

### <span style="color:green"><b>C1. Data Goals</b>:</span>

My approach will include:
<br>&ensp; 1. Back up my data and the process I am following as a copy to my machine and, since this is a manageable dataset, to GitHub using command line and gitbash.
<br>&ensp; 2. Read the data set into Python using Pandas' read_csv command.
<br>&ensp; 3. Evaluate the data struture to better understand input data.
<br>&ensp; 4. Naming the dataset as a the variable "churn_df" and subsequent useful slices of the dataframe as "df".
<br>&ensp; 5. Examine potential misspellings, awkward variable naming & missing data.
<br>&ensp; 6. Find outliers that may create or hide statistical significance using histograms.
<br>&ensp; 7. Imputing records missing data with meaningful measures of central tendency (mean, median or mode) or simply remove outliers that are several standard deviations above the mean.

&emsp; Most relevant to our decision making process is the <b>dependent variable</b> of "Churn" which is binary categorical with only two values, Yes or No.  Churn will be our <b>categorical target variable</b>. 

<br>In cleaning the data, we may discover relevance of the <b>continuous predictor variables</b>: 

* Children
* Income
* Outage_sec_perweek
* Email
* Contacts    
* Yearly_equip_failure
* Tenure (the number of months the customer has stayed with the provider)
* MonthlyCharge
* Bandwidth_GB_Year    
    
<br>&emsp; Likewise, we may discover relevance of the <b>categorical predictor variables</b> (all binary categorical with only two values, Yes or No, except where noted): 

* Techie: Whether the customer considers themselves technically inclined (based on customer questionnaire when they signed up for services) (yes, no)
* Contract: The contract term of the customer (month-to-month, one year, two year)
* Port_modem: Whether the customer has a portable modem (yes, no)
* Tablet: Whether the customer owns a tablet such as iPad, Surface, etc. (yes, no)
* InternetService: Customer’s internet service provider (DSL, fiber optic, None)
* Phone: Whether the customer has a phone service (yes, no)
* Multiple: Whether the customer has multiple lines (yes, no)
* OnlineSecurity: Whether the customer has an online security add-on (yes, no)
* OnlineBackup: Whether the customer has an online backup add-on (yes, no)
* DeviceProtection: Whether the customer has device protection add-on (yes, no)
* TechSupport: Whether the customer has a technical support add-on (yes, no)
* StreamingTV: Whether the customer has streaming TV (yes, no)
* StreamingMovies: Whether the customer has streaming movies (yes, no)
    
<br>&emsp; Finally, <b>discrete ordinal predictor variables</b> from the survey responses from customers regarding various customer service features may be relevant in the decision-making process. &nbsp; In the surveys, customers provided ordinal numerical data by rating 8 customer service factors on a scale of 1 to 8 (1 = most important, 8 = least important): 
    
* Item1: Timely response
* Item2: Timely fixes
* Item3: Timely replacements
* Item4: Reliability
* Item5: Options
* Item6: Respectful response
* Item7: Courteous exchange
* Item8: Evidence of active listening


### <span style="color:green"><b>C2. Summary Statistics</b>:</span>
&emsp; As output by Python pandas dataframe methods below, the dataset consists of 50 original columns & 10,000 records. &nbsp; For purposes of this analysis certain user ID & demographic categorical variables (CaseOrder, Customer_id, Interaction, UID, City, State, County, Zip, Lat, Lng, Population, Area, TimeZone, Job, Marital, PaymentMethod) were removed from the dataframe. &nbsp; Also, binomial Yes/No or Male/Female, variables were encoded to 1/0, respectively. &nbsp; This resulted in 34 remaining numerical variables, including the target variable. &nbsp; The dataset appeared to be sufficiently cleaned leaving no null, NAs or missing data points.
<br> &emsp; Measures of central tendency through histograms & boxplots revealed normal distributions for Monthly_Charge, Outage_sec_perweek & Email. &nbsp; The cleaned dataset no longer retained any outliers. &nbsp; Histograms for Bandwidth_GB_Year & Tenure displayed bimodal distributions, which demonstrated a direct linear relationship with each other in a scatterplot. &nbsp; The average customer was 53 years-old (with a standard deviation of 20 years), had 2 children  (with a standard deviation of 2 kids), an income of 39,806 (with a standard deviation of about 30,000), experienced 10 outage-seconds/week, was marketed to by email 12 times, contacted technical support less than one time, had less than 1 yearly equipment failure, has been with the company for 34.5 months, has a monthly charge of approximately 173 & uses 3,392 GBs/year.

### <span style="color:green"><b>C3. Steps to Prepare Data</b>:</span>

* Import dataset to Python dataframe.
* Rename columns/variables of survey to easily recognizable features (ex: "Item1" to "TimelyResponse").
* Get a description of dataframe, structure (columns & rows) & data types.
* View summary statistics.
* Drop less meaningful identifying (ex: "Customer_id") & demographic columns (ex: zip code) from dataframe.
* Check for records with missing data & impute missing data with meaningful measures of central tendency (mean, median or mode) or simply remove outliers that are several standard deviations above the mean.
* Create dummy variables in order to encode categorical, yes/no data points into 1/0 numerical values.
* View univariate & bivariate visualizations.
* Place "Churn" at end of dataframe
* Finally, the prepared dataset will be extracted & provided as "churn_prepared_log.csv"

In [ ]:
# Increase Jupyter display cell-width
from IPython.core.display import display, HTML
display(HTML("<style>.container { width:75% !important; }</style>"))

In [ ]:
# Standard data science imports
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

# Visualization libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# Statistics packages
import pylab
from pylab import rcParams
import statsmodels.api as sm
import statistics
from scipy import stats

# Scikit-learn
import sklearn
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.metrics import classification_report


# Import chisquare from SciPy.stats
from scipy.stats import chisquare
from scipy.stats import chi2_contingency

# Ignore Warning Code
import warnings
warnings.filterwarnings('ignore')

In [ ]:
# Change color of Matplotlib font
import matplotlib as mpl

COLOR = 'white'
mpl.rcParams['text.color'] = COLOR
mpl.rcParams['axes.labelcolor'] = COLOR
mpl.rcParams['xtick.color'] = COLOR
mpl.rcParams['ytick.color'] = COLOR

In [ ]:
# Load data set into Pandas dataframe
churn_df = pd.read_csv('Data/churn_clean.csv')

# Rename last 8 survey columns for better description of variables
churn_df.rename(columns = {'Item1':'TimelyResponse', 
                    'Item2':'Fixes', 
                     'Item3':'Replacements', 
                     'Item4':'Reliability', 
                     'Item5':'Options', 
                     'Item6':'Respectfulness', 
                     'Item7':'Courteous', 
                     'Item8':'Listening'}, 
          inplace=True)

In [ ]:
# Display Churn dataframe
churn_df

In [ ]:
# List of Dataframe Columns
df = churn_df.columns
print(df)

In [ ]:
# Find number of records and columns of dataset
churn_df.shape

In [ ]:
# Describe Churn dataset statistics
churn_df.describe()

In [ ]:
# Remove less meaningful demographic variables from statistics description
churn_df = churn_df.drop(columns=['CaseOrder', 'Customer_id', 'Interaction', 'UID', 'City', 
                            'State', 'County', 'Zip', 'Lat', 'Lng', 'Population', 
                            'Area', 'TimeZone', 'Job', 'Marital', 'PaymentMethod'])
churn_df.describe()

In [ ]:
# Discover missing data points within dataset
data_nulls = churn_df.isnull().sum()
print(data_nulls)

### Dummy variable data preparation

In [ ]:
churn_df['DummyGender'] = [1 if v == 'Male' else 0 for v in churn_df['Gender']]
churn_df['DummyChurn'] = [1 if v == 'Yes' else 0 for v in churn_df['Churn']] ### If the customer left (churned) they get a '1'
churn_df['DummyTechie'] = [1 if v == 'Yes' else 0 for v in churn_df['Techie']]
churn_df['DummyContract'] = [1 if v == 'Two Year' else 0 for v in churn_df['Contract']]
churn_df['DummyPort_modem'] = [1 if v == 'Yes' else 0 for v in churn_df['Port_modem']]
churn_df['DummyTablet'] = [1 if v == 'Yes' else 0 for v in churn_df['Tablet']]
churn_df['DummyInternetService'] = [1 if v == 'Fiber Optic' else 0 for v in churn_df['InternetService']]
churn_df['DummyPhone'] = [1 if v == 'Yes' else 0 for v in churn_df['Phone']]
churn_df['DummyMultiple'] = [1 if v == 'Yes' else 0 for v in churn_df['Multiple']]
churn_df['DummyOnlineSecurity'] = [1 if v == 'Yes' else 0 for v in churn_df['OnlineSecurity']]
churn_df['DummyOnlineBackup'] = [1 if v == 'Yes' else 0 for v in churn_df['OnlineBackup']]
churn_df['DummyDeviceProtection'] = [1 if v == 'Yes' else 0 for v in churn_df['DeviceProtection']]
churn_df['DummyTechSupport'] = [1 if v == 'Yes' else 0 for v in churn_df['TechSupport']]
churn_df['DummyStreamingTV'] = [1 if v == 'Yes' else 0 for v in churn_df['StreamingTV']]
churn_df['StreamingMovies'] = [1 if v == 'Yes' else 0 for v in churn_df['StreamingMovies']]
churn_df['DummyPaperlessBilling'] = [1 if v == 'Yes' else 0 for v in churn_df['PaperlessBilling']]

In [ ]:
churn_df.head()

In [ ]:
# Drop original categorical features from dataframe
churn_df = churn_df.drop(columns=['Gender', 'Churn', 'Techie', 'Contract', 'Port_modem', 'Tablet', 
                                  'InternetService', 'Phone', 'Multiple', 'OnlineSecurity', 
                                  'OnlineBackup', 'DeviceProtection', 'TechSupport', 
                                  'StreamingTV', 'StreamingMovies', 'PaperlessBilling'])
churn_df.describe()

In [ ]:
df = churn_df.columns
print(df)

In [ ]:
# Move DummyChurn to end of dataset as target
churn_df = churn_df[['Children', 'Age', 'Income', 'Outage_sec_perweek', 'Email', 'Contacts',
       'Yearly_equip_failure', 'Tenure', 'MonthlyCharge', 'Bandwidth_GB_Year', 
        'TimelyResponse', 'Fixes', 'Replacements',
       'Reliability', 'Options', 'Respectfulness', 'Courteous', 'Listening',
       'DummyGender', 'DummyTechie', 'DummyContract',
       'DummyPort_modem', 'DummyTablet', 'DummyInternetService', 'DummyPhone',
       'DummyMultiple', 'DummyOnlineSecurity', 'DummyOnlineBackup',
       'DummyDeviceProtection', 'DummyTechSupport', 'DummyStreamingTV',
       'DummyPaperlessBilling', 'DummyChurn']]

In [ ]:
df = churn_df.columns
print(df)

### <span style="color:green"><b>C4. Visualizations</b>:</span>

In [ ]:
# Visualize missing values in dataset

# Install appropriate library
!pip install missingno

# Importing the libraries
import missingno as msno

# Visualize missing values as a matrix
msno.matrix(churn_df);

## Univariate Statistics

In [ ]:
# Create histograms of contiuous variables
churn_df[['Children', 'Age', 'Income', 'Outage_sec_perweek', 'Email', 
          'Contacts', 'Yearly_equip_failure', 'Tenure', 'MonthlyCharge', 
          'Bandwidth_GB_Year']].hist()
plt.savefig('churn_pyplot.jpg')
plt.tight_layout()

In [ ]:
# Create Seaborn boxplots for continuous variables
sns.boxplot('Tenure', data = churn_df)
plt.show()

In [ ]:
sns.boxplot('MonthlyCharge', data = churn_df)
plt.show()

In [ ]:
sns.boxplot('Bandwidth_GB_Year', data = churn_df)
plt.show()

## Anomalies
### &emsp; It appears that anomolies have been removed from the supplied dataset, churn_clean.csv. &nbsp; There are no remaining outliers.

## Bivariate Statistics

In [ ]:
# Run scatterplots to show direct or inverse relationships between target & independent variables
sns.scatterplot(x=churn_df['Children'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Age'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Income'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['DummyGender'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Outage_sec_perweek'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Email'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Contacts'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Yearly_equip_failure'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['DummyTechie'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Tenure'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['MonthlyCharge'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Bandwidth_GB_Year'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['TimelyResponse'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Fixes'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Replacements'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Reliability'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Options'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Respectfulness'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Courteous'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['Listening'], y=churn_df['DummyChurn'], color='red')
plt.show();

In [ ]:
sns.scatterplot(x=churn_df['MonthlyCharge'], y=churn_df['Outage_sec_perweek'], color='red')
plt.show();

## Scatterplot Summary
### &emsp; These scatterplots suggest no correlation between a customer churning (Churn = 1) & any of our continous user data points or categorical responses to survey data points.

### <span style="color:green"><b>C5. Prepared Dataset</b>:</span>

In [ ]:
# Extract Clean dataset
churn_df.to_csv('churn_prepared_log.csv')

### <span style="color:green"><b>D1. Initial Model</b></span>

In [ ]:
"""Develop the initial estimated regression equation that could be used to predict the probability of customer churn, given the only continuous variables"""
churn_df = pd.read_csv('churn_prepared_log.csv')
churn_df['intercept'] = 1
churn_df = pd.get_dummies(churn_df, drop_first=True)

churn_logit_model = sm.Logit(churn_df['DummyChurn'], churn_df[['Children', 'Age', 
                                                               'Income',
                                                               'Outage_sec_perweek', 
                                                               'Email', 'Contacts',
                                                               'Yearly_equip_failure', 
                                                               'Tenure', 'MonthlyCharge',
                                                               'Bandwidth_GB_Year',
                                                               'TimelyResponse', 'Fixes', 
                                                               'Replacements', 'Reliability', 
                                                               'Options', 'Respectfulness', 
                                                               'Courteous', 'Listening', 
                                                               'intercept']]).fit()
print(churn_logit_model.summary())

## Dummy Variables
### Now, we will run a model including all encoded categorical dummy variables.

In [ ]:
""""Model including all dummy variables"""
churn_df = pd.read_csv('data/churn_prepared_log.csv')
churn_df['intercept'] = 1
churn_df = pd.get_dummies(churn_df, drop_first=True)

churn_logit_model2 = sm.Logit(churn_df['DummyChurn'], churn_df[['Children', 'Age', 
                                                               'Income', 'Outage_sec_perweek', 
                                                               'Email', 'Contacts',
                                                               'Yearly_equip_failure', 
                                                               'DummyTechie', 'DummyContract', 
                                                               'DummyPort_modem', 'DummyTablet', 
                                                               'DummyInternetService', 'DummyPhone', 
                                                               'DummyMultiple', 'DummyOnlineSecurity', 
                                                               'DummyOnlineBackup', 'DummyDeviceProtection', 
                                                               'DummyTechSupport', 'DummyStreamingTV', 
                                                               'DummyPaperlessBilling',
                                                               'Tenure', 'MonthlyCharge', 'Bandwidth_GB_Year', 
                                                               'TimelyResponse', 'Fixes', 
                                                               'Replacements', 'Reliability', 
                                                               'Options', 'Respectfulness', 
                                                               'Courteous', 'Listening', 
                                                               'intercept']]).fit()
print(churn_logit_model2.summary())

## Early Model Comparison
### Following the second run of our MLE model, our pseudo R went up from 0.4473 to 0.5296 as we added in our categorical dummy variables to our continuous variables. &nbsp; We will take that as a good sign that some of the explanation of our variance is within the categorical data points. &nbsp; We will use those 31 variables as our initial regression equation.

### Initial Multiple Linear Regression Model
With <b><i>31</i></b> independent variables (18 continuous & 13 categorical): 
<br><br>&emsp;<span style="color:gold"><b><i>y</i></b> &nbsp; = &nbsp; -5.8583 + (-0.0395 * Children) + (0.0069 * Age) + (1.199e-07 * Income) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.0020 * Outage_sec_perweek) + (-0.0015 * Email) + (0.0301 * Contacts) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.0308 * Yearly_equip_failure) + (0.7956 * DummyTechie) + (-2.295 * DummyContract) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (0.161 * DummyPort_modem) + (-0.0796 * DummyTablet) + (-1.4252 * DummyInternetService) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.3157 * DummyPhone) + (-0.2908 * DummyMultiple) + (-0.3280 * DummyOnlineSecurity) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.5125 * DummyOnlineBackup) + (-0.41 * DummyDeviceProtection) + (-0.3461 * DummyTechSupport) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (0.0311 * DummyStreamingTV) + (0.1126 * DummyPaperlessBilling) + (-0.2043 * Tenure) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (0.0461 * MonthlyCharge) + (0.0013 * Bandwidth_GB_Year) + (-0.0167 * TimelyResponse) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (0.0143 * Fixes) + (-0.0158 * Replacements) + (-0.025 * Reliability) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.0341 * Options) + (-0.0309 * Respectfulness) + (0.0047 * Courteous) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.009 * Listening)</span>

### <span style="color:green"><b>D2. Justification of Model Reduction</b></span>
&emsp; Based on the above MLE model we created, we have a pseudo R value = 0.5296, which is clearly not very good for the variance of our model. &nbsp; Also, coefficients on the above model are very low (less than 0.5) with the exception of variables DummyTechie, DummyContract, DummyInternetService & DummyOnlineBackup. &nbsp; Those variables also have p-values less than 0.000 & appear, therefore, significant.  
&emsp; Subsequently, let us choose a p-value of 0.05 & include all variables with p-values &#8804; 0.05. &nbsp; We will remove any predictor variable with a p-value greater than 0.05 as not statistically significant to our model.

<br>Our next MLE run will include the <b>continuous predictor variables</b>: 

* Age  
* Tenure
* MonthlyCharge
* Bandwidth_GB_Year

And, <b>categorical predictor variables</b>:

* DummyTechie 
* DummyContract
* DummyPort_modem
* DummyInternetService
* DummyPhone 
* DummyMultiple
* DummyOnlineSecurity
* DummyOnlineBackup
* DummyDeviceProtection
* DummyTechSupport

We will run that reduced number of predictor variables against our DummyChurn dependent variable in another MLE model.

### <span style="color:green"><b>D3. Reduced Multiple Regression Model</b></span>

In [ ]:
# Run reduced OLS multiple regression
churn_df['intercept'] = 1
churn_logit_model_reduced = sm.Logit(churn_df['DummyChurn'], 
                                     churn_df[['Children', 'Age', 'DummyTechie', 'DummyContract', 'DummyPort_modem', 
                                               'DummyInternetService', 'DummyPhone','DummyMultiple', 
                                               'DummyOnlineSecurity', 'DummyOnlineBackup', 'DummyDeviceProtection', 
                                               'DummyTechSupport', 'Tenure', 'MonthlyCharge', 'Bandwidth_GB_Year', 
                                               'intercept']]).fit()
print(churn_logit_model_reduced.summary())

Optimization terminated successfully.
         Current function value: 0.272362
         Iterations 8
                           Logit Regression Results                           
Dep. Variable:             DummyChurn   No. Observations:                10000
Model:                          Logit   Df Residuals:                     9984
Method:                           MLE   Df Model:                           15
Date:                Wed, 21 Jul 2021   Pseudo R-squ.:                  0.5290
Time:                        10:31:53   Log-Likelihood:                -2723.6
converged:                       True   LL-Null:                       -5782.2
Covariance Type:            nonrobust   LLR p-value:                     0.000
                            coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------
Children                 -0.0391      0.018     -2.221      0.026      -0.074      -0.

### Reduced Logistic Regression Model
With <b>15</b> independent variables (5 continuous & 10 categorical):

<br><br>&emsp;<span style="color:gold"><b><i>y</i></b> &nbsp; = &nbsp; -6.1973 + (-0.0391 * Children) + (0.0070 * Age) + 
(0.7970 * DummyTechie) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-2.2895 * DummyContract) + (0.1598 * DummyPort_modem) + (-1.4240 * DummyInternetService) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.3193 * DummyPhone) + (-0.2964 * DummyMultiple) + (-0.3303 * DummyOnlineSecurity) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.5146 * DummyOnlineBackup) + (-0.41 * DummyDeviceProtection) + (-0.3461 * DummyTechSupport) + 
<br> &emsp; &emsp; &emsp; &emsp; &emsp; &emsp; &nbsp; (-0.2049 * Tenure) + (0.0463 * MonthlyCharge) + (0.0013 * Bandwidth_GB_Year)</span>

### <span style="color:green"><b>E1. Model Comparison</b></span>
The second model still explains 52% of variance, as demonstrated by the pseudo R, even though we have reduced the number of variables in half (from 31 to 15).  We have suggested an alpha threshold of 0.05 to retain predictor variables.  We can see that, as Churn = 1 & that our majority of our dummy variables (which are additional services that a customer may add on to their contract) have negative values.  
&emsp; What is important to decision-makers & marketers is that those inverse relationships suggest that as a customer subscribes to more services that the company provided, an additional port modem or online backup for example, they are less likely to churn & leave the company.  Cleary, it is in the best interest of retaining customers to provide them with more services & improve their experience with the company by helping customers understand all the services that are available to them as a subscriber, not simply mobile phone service.

### Confusion Matrix

In [ ]:
# Import the prepared dataset
dataset = pd.read_csv('data/churn_prepared_log.csv')
X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, -1].values

In [ ]:
# Split the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [ ]:
# Training the Logistic Regression model on the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

In [ ]:
# Predict the Test set results
y_pred = classifier.predict(X_test)

In [ ]:
# Make the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

In [ ]:
## Compute the accuracy with k-Fold Cross Validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(accuracies.std()*100))

In [ ]:
y_predict_test = classifier.predict(X_test)
cm2 = confusion_matrix(y_test, y_predict_test)
sns.heatmap(cm2, annot=True)

### Classification Report

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict_test))

### <span style="color:green"><b>E2. Output & Calculations</b></span>
Calculations & code output above.

### <span style="color:green"><b>E3. Code</b></span>
All code for analysis include above.

### <span style="color:green"><b>F1. Results</b></span>
The final multiple regression equation with <b>15</b> predictor variables: 
<br><br>&emsp;<span style="color:gold">y = -6.1973 - 0.0391 * Children + 0.0070 * Age + 0.7970 * DummyTechie - 2.2895 * DummyContract + 0.1598 * DummyPort_modem - 1.4240 * DummyInternetService - 0.3193 * DummyPhone - 0.2964 * DummyMultiple - 0.3303 * DummyOnlineSecurity - 0.5146 * DummyOnlineBackup - 0.4075 * DummyDeviceProtection - 0.3555 * DummyTechSupport - 0.2049 * Tenure + 0.0463 * MonthlyCharge + 0.0013 * Bandwidth_GB_Year</span>

### <span style="color:green"><b>F2. Recommendations</b></span>
It is critical that decision-makers & marketers understand that there is inverse relationship between our target variable of Churn & several of our predictor variables.  This suggests that as a customer subscribes to more services that the company provided, an additional port modem or online backup for example, they are less likely to leave the company. Clearly, it is the best interest of retaining customers to provide them with more services & improve their experience with the company by helping customers understand all the services that are available to them as a subscriber, not simple mobile phone service.  Given the negative coefficients of additional services, we suggest additional marketing efforts for contracts & internet services as those with contract appear less likely to leave the company.  
&emsp; Also, with such a direct linear relationship between bandwidth used yearly & tenure with the telecom company it makes sense to suggest the company do everything within marketing & customer service capability to retain the customers gained as the longer they stay with the company the more bandwidth they tend to use.  This would include making sure that fixes to customer problems are prompt & that the equipment provided is high quality to avoid fewer replacements of equipment. 

### <span style="color:green"><b>G. Video</b></span>
<span style="color:red">link</span>

### <span style="color:green">H. Sources for Third-Party Code</span>

GeeksForGeeks. &ensp; (2019, July 4). &ensp; <i>Python | Visualize missing values (NaN) values using Missingno Library</i>. &ensp; GeeksForGeeks. &ensp; https://www.geeksforgeeks.org/python-visualize-missing-values-nan-values-using-missingno-library/

### <span style="color:green">I. Sources</span>

CBTNuggets. &ensp; (2018, September 20). &ensp; <i>Why Data Scientists Love Python</i>. &ensp; https://www.cbtnuggets.com/blog/technology/data/why-data-scientists-love-python

<br> Massaron, L. & Boschetti, A. &ensp; (2016). &ensp; <i>Regression Analysis with Python</i>. &ensp; Packt Publishing.

In [ ]:
!wget -nc https://raw.githubusercontent.com/brpy/colab-pdf/master/colab_pdf.py
from colab_pdf import colab_pdf
colab_pdf('D208_Performance_Assessment_NBM2_Task_2.ipynb')